runtime/array.c

```diff
@@ -327,43 +315,12 @@ CAMLprim value caml_floatarray_blit(value a1, value ofs1, value a2, value ofs2,
 CAMLprim value caml_array_blit(value a1, value ofs1, value a2, value ofs2,
                                value n)
 {
-  value * src, * dst;
-  intnat count;
-
 #ifdef FLAT_FLOAT_ARRAY
   if (Tag_val(a2) == Double_array_tag)
     return caml_floatarray_blit(a1, ofs1, a2, ofs2, n);
 #endif
   CAMLassert (Tag_val(a2) != Double_array_tag);
-  if (Is_young(a2)) {
-    /* Arrays of values, destination is in young generation.
-       Here too we can do a direct copy since this cannot create
-       old-to-young pointers, nor mess up with the incremental major GC.
-       Again, memmove takes care of overlap. */
-    memmove(&Field(a2, Long_val(ofs2)),
-            &Field(a1, Long_val(ofs1)),
-            Long_val(n) * sizeof(value));
-    return Val_unit;
-  }
-  /* Array of values, destination is in old generation.
-     We must use caml_modify.  */
-  count = Long_val(n);
-  if (a1 == a2 && Long_val(ofs1) < Long_val(ofs2)) {
-    /* Copy in descending order */
-    for (dst = &Field(a2, Long_val(ofs2) + count - 1),
-           src = &Field(a1, Long_val(ofs1) + count - 1);
-         count > 0;
-         count--, src--, dst--) {
-      caml_modify(dst, *src);
-    }
-  } else {
-    /* Copy in ascending order */
-    for (dst = &Field(a2, Long_val(ofs2)), src = &Field(a1, Long_val(ofs1));
-         count > 0;
-         count--, src++, dst++) {
-      caml_modify(dst, *src);
-    }
-  }
+  caml_blit_fields(a1, Long_val(ofs1), a2, Long_val(ofs2), Long_val(n));
```

Comment:
Note that we can't use `memmove` here since they may not do atomic word writes. For instance, they may copy a byte at a time. The implementation of `caml_blit_fields` in multicore does not use `memmove` and `memcpy`, and copies word at a time.

Reply:
In order to make the diff simpler, it may be useful to implement a word-sized memcpy and memmove, and use that in these files.

Reply:
`caml_blit_fields` is incorrect in general: if the tag is >= No_scan_tag, it does not contain well-formed values and it's wrong to use `caml_modify` on them.  I would recommend NOT defining this function in memory.c and NOT exporting it.  A `static` version can be defined in `array.c`, where we know we'll be calling it on zero-tagged blocks.

Reply:
Alternate recommendation (as I wrote elsewhere): revert to the array.c code currently in trunk.

Reply:
I agree that starting from the trunk version of `array.c` is the best way to go.

Reply:
I've reverted the array.c changes and have made a PR: ocaml-multicore/ocaml-multicore#822. It would benefit from a quick review.